### ISSUE TITLE
Training with NCC v/s MSE

### ISSUE THREAD

**Issue by varsha-raveendran** [cwmok/LapIRN#25]
Hi @cwmok,

As I mentioned [here](https://github.com/cwmok/LapIRN/issues/24), I'm comparing methods using masked inputs vs. non-masked inputs. This is done as part of my thesis. I'm using the NLST dataset from L2R 2022.
An example training image after clipping: 
![image](https://github.com/cwmok/LapIRN/assets/67820303/de57d4b5-4255-4f40-8a56-bfc0639c21ef)

I normalize and downsample the images for training. With non-masked inputs (i.e background intensity is not equal to zero), the training loss increases and the dice score is low (<0.5). 
![image](https://github.com/cwmok/LapIRN/assets/67820303/4bb35eb9-d4b4-41a2-ac69-b834dc111664)

I observed a similar behaviour with Voxelmorph with NCC as the intensity loss. I then tried MSE for Voxelmorph and the loss curves alongwith the results were much better. 

1. I saw that you mention having the background intensity set to zero while using NCC loss in other issues. I understand that this is different from masking the lungs/RoI or am I missing something?
2. I'm not sure how to use LapIRN in this scenario where I don't mask the inputs. Could the issue I'm facing be due to the behaviour of NCC loss or something else? 

I'm relatively new to this field and would appreciate any help. 

Best,
Varsha

**Comment by cwmok** [comment]
Hi @varsha-raveendran,

>> I saw that you mention having the background intensity set to zero while using NCC loss in other issues. I understand that this is different from masking the lungs/RoI or am I missing something?

To my understanding, I think the problem is because the number of background voxels and the air within the lung region dominated the NCC calculation when the background intensity is not set to zero/not using lung lobe mask to mask the similarity. You can verify this assumption by plotting the NCC map of the fixed and moving images. The NCC map will have 0 value for background intensity (automatically masking the background region if the background intensity is set to zero).

In that case (background intensity is not zero and not masked in NCC calculation), the background will become the region of interest to the registration algorithm and the training becomes unstable. It can be solved by 1) using a sparser deformation model which less degree of freedom, 2) changing the similarity function or 3) (optimal) using lung lobe mask to mask out the irrelevant region in the similarity metric.

>> I'm not sure how to use LapIRN in this scenario where I don't mask the inputs. Could the issue I'm facing be due to the behaviour of NCC loss or something else?

I think so. Changing to another loss function will work. But I suggest using the provided lung lobe mask of the fixed image to mask the similarity metric during training. (In inference phase, no mask is needed.)

**Comment by varsha-raveendran** [comment]
Thank you, I will take a look at  cost function masking.

**Comment by cwmok** [comment]
Hi @varsha-raveendran,

Recently, I faced a similar problem as yours. I found that the eps value in the NCC function is important. If you are facing training instability with NCC, you may try to use a smaller value of eps, i.e., NCC(win=5, eps=1e-4). 

Hope it helps.

**Comment by varsha-raveendran** [comment]
@cwmok Thank you! I started training with eps=1e-4 now and the training seems stable for level 1 and 2. I will update here regarding the results once the training completes. 

Additionally, I tried the LNCC loss from Monai and it also gave me a stable loss curve.

I observed a weird behaviour with a couple of experiments. It would be good to have your thoughts regarding this and suggestions on how I could debug:

1. I thought with masked inputs the performance would be better than non-masked inputs. However, the TRE and Dice score for non-masked inputs are better than masked inputs. With Monai's LNCC loss, I get TRE of ~4 for non-masked. However, the TRE is > 7 for masked. I'm not sure where I'm going wrong or if my assumption is wrong. (Same behaviour with Voxelmorph as well). I'm not sure if this is the correct way to analyse - I plotted the feature map of the last-but one layer in UNet from Voxelmorph for both the models - non-masked and masked. These are the plots: First one is for non-masked and 2nd for masked. 
![image](https://github.com/cwmok/LapIRN/assets/67820303/9a1dcf8b-bbd5-4291-af9a-af5e76e3f103)

![image](https://github.com/cwmok/LapIRN/assets/67820303/aa25c4a6-c7bb-44b7-81da-ea3dc7c68bb0)

2. Similarly, I added dice loss to the total loss. I see a similar behaviour - the non-masked performs better than masked + dice. I tried different weights for the dice loss term but didn't see much difference.  Here's how I added the weakly supervised loop: I used a different (from the one used for image warping) STN layer to warp the mask at each level. In the train loop, I then calculated the dice loss using Voxelmorph's implementation. Finally, added the dice loss to the total loss term. 

3. Could you also give me any pointers on how to choose lambda for regularization loss? I tried 0.001, 0.01, 0.1, 1, 3 but I'm not sure how to choose a value for a particular dataset. 



Best,
Varsha

**Comment by cwmok** [comment]
Hi @varsha-raveendran,

Thanks for sharing your solution. I will also try out the LNCC loss from the Monai.

Regarding your questions, I summarize my response below:
1. I did achieve a better TRE with the masked similarity measure. Note that I didn't directly mask the input image. Instead, I masked the NCC value using the mask, i.e.
````
cc = cross * cross / (I_var * J_var + self.eps)

# return negative cc.
if mask is None:
    return -1.0 * torch.mean(cc)
else:
    return -1.0 * (torch.sum(cc*mask)/torch.sum(mask))
````
I also dilated the lung lobe mask a little bit so that the boundary of the lung lobe is included in the mask.

2. Empirically, I found the DICE loss will hinder the TRE if added to the loss function. Therefore, I don't recommend using DICE loss for this task.

3. That's actually the problem I tried to solve in my [MICCAI 2021 paper](https://arxiv.org/abs/2106.12673). Instead of trying different lambda one by one, I redesigned the learning paradigm of LapIRN and forced it to learn the effect of a set of lambda with one model. Then, the best lambda can be easily found during the inference phase using grid search.

In general, lambda is set to 1-5 in LapIRN. The actual value will depend on the dataset and the resolution of the image.

**Comment by varsha-raveendran** [comment]
Hi @cwmok,

Thank you :) I will mask the NCC value and check the Conditional LapIRN repo for lambda.

**Comment by varsha-raveendran** [comment]
Hi @cwmok,

Do you also use masks for the multiscale NCC loss? If yes, do you downsample the masks to different scales and then use it to mask NCC?

**Comment by varsha-raveendran** [comment]
Hi @cwmok,

Do you also use masks for the multiscale NCC loss? If yes, do you downsample the masks to different scales and then use it to mask NCC?

**Comment by cwmok** [comment]
Hi @varsha-raveendran,

I tried both approaches for Lung registration. I found that using multiscale NCC loss is inferior to a single-scale NCC loss in lung registration. I use NCC with windows size = 5 for lung registration. But I use multiscale NCC loss for Brain MR (OASIS) registration to achieve the best results.

When multiscale NCC loss is used, I downsampled the masks to different scales using 3D average pooling.

**Comment by varsha-raveendran** [comment]
Thank you for the quick response :) Any idea on why multiscale loss performs inferior to single-scale?

**Comment by cwmok** [comment]
Hi @varsha-raveendran,

I am not sure. For lung registration, the choice of smoothness regularization and transformation model also matter. How to achieve a good balance between the smoothness of the deformation field and the registration accuracy remains an open issue.

**Comment by varsha-raveendran** [comment]
Understood :) Thanks again!
